Primary sort order for IDoValueMigrationHandler (#231)
BSI-Business-Systems-Integration-AG/org.eclipse.scout.docs: docs/modules/migration/partials/migration-guide-24.2.adoc

|PermissionId
|id
|===

[[value-migration-primary-sort-order]]
== Primary sort order for IDoValueMigrationHandler

Untyped data object value migration handlers need to be applied before regular value migrations because they may convert typed values to the expected target type.

> A typical use case is conversion from `UnknownId` to a typed target ID in case of renaming a type ID name.

Reviewer: ```suggestion
A typical use case is the conversion from `UnknownId` to a typed target ID in case of renaming a type ID name.
```

> Classes implementing `IDoValueMigrationHandler` need to implement the new method `IDoValueMigrationHandler.primarySortOrder()`.

Reviewer 2: or use AbstractDoValueMigrationHandler as baseclass which defines DEFAULT_PRIMARY_SORT_ORDER as default

> Classes implementing `IDoValueMigrationHandler` need to implement the new method `IDoValueMigrationHandler.primarySortOrder()`.

Reviewer: ```suggestion
Classes implementing `IDoValueMigrationHandler` directly need to implement the new method `IDoValueMigrationHandler.primarySortOrder()`.
```

> Existing value migration handlers typically should return `IDoValueMigrationHandler.DEFAULT_PRIMARY_SORT_ORDER` or use `AbstractDoValueMigrationHandler` as baseclass which defines `DEFAULT_PRIMARY_SORT_ORDER` by default.

Reviewer: ```suggestion
It's highly recommended to extend `AbstractDoValueMigrationHandler` which defines `DEFAULT_PRIMARY_SORT_ORDER` by default instead of implementing only the interface.
```

The primary sort order can be used to group different types of value migrations.

> Regular value migrations (i.e. sub-classes of `AbstractDoValueMigrationHandler` use `IDoValueMigrationHandler.DEFAULT_PRIMARY_SORT_ORDER` and are applied after untyped value migrations (i.e. sub-classes of `AbstractDoValueUntypedMigrationHandler`) which use `IDoValueMigrationHandler.UNTYPED_PRIMARY_SORT_ORDER`.

Reviewer: ```suggestion
Regular value migrations (i.e. sub-classes of `AbstractDoValueMigrationHandler` use `IDoValueMigrationHandler.DEFAULT_PRIMARY_SORT_ORDER`) and are applied after untyped value migrations (i.e. sub-classes of `AbstractDoValueUntypedMigrationHandler`) which use `IDoValueMigrationHandler.UNTYPED_PRIMARY_SORT_ORDER`.
```
